[mlir][emitc] Deferred emission as expressions

aniragil opened the pull request:
The translator currently implements deferred emission for certain ops. Like expressions, these ops are emitted as part of their users but unlike expressions, this is mandatory. Besides complicating the code with a second inlining mechanism, deferred emission's inlining is limited as it's not recursive.

This patch extends EmitC's expressions to deferred emission ops by (a) marking them as CExpressions, (b) extending expression interface to mark ops as always-inline and (c) support inlining of always-inline CExpressions even when not packed of an `emitc.expression` op, retaining current behavior.

aniragil committed:
[mlir][emitc] Deferred emission as expressions

The translator currently implements deferred emission for certain ops.
Like expressions, these ops are emitted as part of their users but
unlike expressions, this is mandatory. Besides complicating the code
with a second inlining mechanism, deferred emission's inlining is
limited as it's not recursive.

This patch extends EmitC's expressions to support the always-inline
semantics required by the deferred emission ops by adding an
`always-inline` property to the CExpression interface, which forces
unconditional inlining whether such ops appear within an
`emitc.expression` op or not. With deferred emission ops converted
to always-inline CExpressions, current behavior is retained.

aniragil commented:
@Vladislave0-0 this patch also addresses literals, per [your patch](https://github.com/llvm/llvm-project/pull/155582).

marbre commented on mlir/lib/Target/Cpp/TranslateToCpp.cpp:
```diff
@@ -1288,44 +1371,14 @@ CppEmitter::CppEmitter(raw_ostream &os, bool declareVariablesAtTop,
   labelInScopeCount.push(0);
 }
 
-std::string CppEmitter::getSubscriptName(emitc::SubscriptOp op) {
-  std::string out;
-  llvm::raw_string_ostream ss(out);
-  ss << getOrCreateName(op.getValue());
-  for (auto index : op.getIndices()) {
-    ss << "[" << getOrCreateName(index) << "]";
-  }
-  return out;
-}
-
-std::string CppEmitter::createMemberAccess(emitc::MemberOp op) {
-  std::string out;
-  llvm::raw_string_ostream ss(out);
-  ss << getOrCreateName(op.getOperand());
-  ss << "." << op.getMember();
-  return out;
-}
-
-std::string CppEmitter::createMemberAccess(emitc::MemberOfPtrOp op) {
-  std::string out;
-  llvm::raw_string_ostream ss(out);
-  ss << getOrCreateName(op.getOperand());
-  ss << "->" << op.getMember();
-  return out;
-}
-
-void CppEmitter::cacheDeferredOpResult(Value value, StringRef str) {
-  if (!valueMapper.count(value))
-    valueMapper.insert(value, str.str());
+void CppEmitter::setName(Value val, StringRef name) {
```
I don't have a strong preferrence with regards to bring back the name `cacheDeferredOpResult` but maybe we find something more descriptive than `setName`?

marbre reviewed (commented):
It would be good to get some more eyes on this as it is changes quite some lines :)

aniragil commented on mlir/test/Target/Cpp/member.mlir:
```diff
@@ -31,9 +31,9 @@ func.func @member(%arg0: !emitc.opaque<"mystruct">, %arg1: i32, %arg2: index) {
 // CPP-DEFAULT-NEXT: [[V4]] = [[V3]];
-// CPP-DEFAULT-NEXT: int32_t [[V5:[^ ]*]] = [[V2]].c[[[Index]]];
+// CPP-DEFAULT-NEXT: int32_t [[V5:[^ ]*]] = ([[V2]].c)[[[Index]]];
 // CPP-DEFAULT-NEXT: [[V4]] = [[V5]];
-// CPP-DEFAULT-NEXT: [[V2]].d[[[Index]]] = [[V1]];
+// CPP-DEFAULT-NEXT: ([[V2]].d)[[[Index]]] = [[V1]];
```
It's a side effect: When emitting expressions we currently don't omit parentheses when operator precedence is equal and associativity is left-to-right, as is the case for `.` and `[]`.

aniragil commented:
> It would be good to get some more eyes on this as it is changes quite some lines :)

@simon-camp, @mgehre-amd could you also take a look?

aniragil commented:
Patch rebased in order to minimize diff (following the recent #167532).
Also removed the new `setName()` method (was not needed in the first place).

aniragil commented:
Patch rebased to minimize diff following #169978.

aniragil commented:
@marbre, @jacquesguan, any more comments for this change?
[Supporting changes within the original patch have been committed separately in #167532, #168267, #168361 and #169978. The patch should now contain only the gist of this PR, i.e. turns deferred-emission ops into always-inline CExpressions, add always-inline logic to `shouldBeInlined()` and removes the deferred emission mechanism].

aniragil commented:
@banach-space , @Jimmy2027 could you take a look?

banach-space commented:
> @banach-space , @Jimmy2027 could you take a look?

I will try to allocate some cycles for this over the coming days 👍🏻 Feel free to land if somebody else approves in the meantime.

Jimmy2027 commented on mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:
```diff
@@ -493,6 +494,15 @@ def EmitC_DereferenceOp : EmitC_Op<"dereference", [
   }];
   let hasVerifier = 1;
+
+  let extraClassDeclaration = [{
```
would a mixin `AlwaysInlineCExpression` make sense here? 

to set this in a single place:
```
      bool hasSideEffects() { return false; }
      bool alwaysInline()   { return true;  }

```

Something like `EmitC_UnaryOp` is doing

aniragil replied:
> would a mixin `AlwaysInlineCExpression` make sense here?
> 
> to set this in a single place:
> 
> ```
>       bool hasSideEffects() { return false; }
>       bool alwaysInline()   { return true;  }
> ```
> 
> Something like `EmitC_UnaryOp` is doing

Hmmm ... good question.
Technically (even though it currently follows) we don't enforce no-side-effects on always-inline ops: as its a translation directive, it might be useful for ops that do have side effects and were placed such that the side effect is movable to the user. So we could have such a convenience trait, but it won't necessarily be used in all cases (which might be misleading).
There's also a small subtlety with LiteralOp compared to the others - it is always inlined by design, not by C's limitation. It doesn't make a difference at this point, but if we add C++ reference support, the others would actually become "conditionally-must-be-inlined", depending on whether we emit C or C++. It then might make sense to have different traits for these cases.
I can add a concrete implementation in this PR, or merge it as-is and open a new PR for this discussion. WDYT?

Jimmy2027 replied:
yeah sounds like this probably warrants its own PR, so let's get this in as is first

Jimmy2027 commented on mlir/lib/Target/Cpp/TranslateToCpp.cpp:
```diff
@@ -1691,6 +1707,12 @@ LogicalResult CppEmitter::emitOperand(Value value, bool isInBrackets) {
+          dyn_cast_if_present<CExpressionInterface>(value.getDefiningOp())) {
+    if (cExpression.alwaysInline())
+      return emitExpression(value.getDefiningOp());
+  }
```
```suggestion
  if (Operation *def = value.getDefiningOp(); def && shouldBeInlined(def))
    return emitExpression(def);
```

shouldBeInlined already encodes "alwaysInline CExpression or inlinable ExpressionOp"

aniragil replied:
> shouldBeInlined already encodes "alwaysInline CExpression or inlinable ExpressionOp"

Good catch!

Jimmy2027 reviewed (commented):
lgtm, clear improvement

aniragil committed:
Implement review comment

In emitOperand it's enough to call shouldBeInlined() on the defining op
of the operand value (if there is one) as it already checks for these
two cases for calling emitExpression.

banach-space reviewed (approved):
Overall, this looks good to me, but it goes deep enough that I would not want to pretend I follow every detail. This is probably one for the EmitC experts :)

Approving on the basis that you are one of the maintainers and already have a +1 from an expert user.

aniragil commented:
> Overall, this looks good to me, but it goes deep enough that I would not want to pretend I follow every detail. This is probably one for the EmitC experts :)
> 
> Approving on the basis that you are one of the maintainers and already have a +1 from an expert user.

Thanks @banach-space!

Jimmy2027 reviewed (commented):
don't have permissions to approve but I think this can go in :)

aniragil commented:
> don't have permissions to approve but I think this can go in :)

Thanks @Jimmy2027!
